webapi/Controllers/ChatHistoryController.cs

```
@@ -240,4 +241,78 @@ public async Task<IActionResult> GetAllChatSessionsAsync(string userId)
+        catch (KeyNotFoundException)
+        {
+            return this.NotFound($"No chat session found for chat id '{chatId}'.");
+        }
+
+        // Delete message and broadcast update to all participants.
+        await this._sessionRepository.DeleteAsync(chatToDelete);
+        await messageRelayHubContext.Clients.Group(chatId).SendAsync(ChatDeletedClientCall, chatId, userId);
+
+        // Create and store the tasks for deleting all users tied to the chat.
+        var participants = await this._participantRepository.FindByChatIdAsync(chatId);
+        var participantsTasks = new List<Task>();
+        foreach (var participant in participants)
+        {
+            participantsTasks.Add(this._participantRepository.DeleteAsync(participant));
+        }
+
+        // Create and store the tasks for deleting chat messages.
+        var messages = await this._messageRepository.FindByChatIdAsync(chatId);
+        var messageTasks = new List<Task>();
+        foreach (var message in messages)
+        {
+            messageTasks.Add(this._messageRepository.DeleteAsync(message));
+        }
+
+        // Create and store the tasks for deleting memory sources.
+        var sources = await this._sourceRepository.FindByChatIdAsync(chatId, false);
+        var sourceTasks = new List<Task>();
+        foreach (var source in sources)
+        {
+            sourceTasks.Add(this._sourceRepository.DeleteAsync(source));
+        }
+
+        // Await all the tasks in parallel and handle the exceptions
+        var cleanupTasks = participantsTasks.Concat(messageTasks).Concat(sourceTasks);
+        await Task.WhenAll(cleanupTasks);
+
+        // Iterate over the tasks and check their status and exception
+        foreach (var task in cleanupTasks)
+        {
+            if (task.IsFaulted && task.Exception != null)
+            {
+                this._logger.LogInformation("Failed to delete an entity of chat {0}: {1}", chatId, task.Exception.Message);
+            }
```

We should return a 500 error in this case so that the user can retry the delete.

---

Returning a 500 now